pkg/apiserver/rest/dualwriter.go

```diff
@@ -117,3 +122,50 @@ func (u *updateWrapper) Preconditions() *metav1.Preconditions {
 func (u *updateWrapper) UpdatedObject(ctx context.Context, oldObj runtime.Object) (newObj runtime.Object, err error) {
 	return u.updated, nil
 }
+
+func SetDualWritingMode(
+	kvs *kvstore.NamespacedKVStore,
+	features featuremgmt.FeatureToggles,
+	entity string,
+	legacy LegacyStorage,
+	storage Storage,
+) (DualWriter, error) {
+	toMode := map[string]DualWriterMode{
+		"1": Mode1,
+		"2": Mode2,
+		"3": Mode3,
+		"4": Mode4,
+	}
+	errDualWriterSetCurrentMode := errors.New("failed to set current dual writing mode")
+
+	// Use entity name as key
+	m, ok, err := kvs.Get(context.Background(), entity)
+	if err != nil {
+		return nil, errors.New("failed to fetch current dual writing mode")
+	}
+	if !ok {
+		// Default to mode 1
+		m = fmt.Sprint(Mode1)
+		err := kvs.Set(context.Background(), entity, m)
+		if err != nil {
+			return nil, errDualWriterSetCurrentMode
+		}
+	}
+
+	currentMode := toMode[m]
+
+	// Desired mode is 2 and current mode is 1
+	if features.IsEnabledGlobally(featuremgmt.FlagDualWritePlaylistsMode2) && (currentMode == Mode1) {
+		// This is where we go through the different gates to allow the instance to migrate from mode 1 to mode 2.
+		// There are none between mode 1 and mode 2
+		currentMode = Mode2
+
+		err := kvs.Set(context.Background(), entity, fmt.Sprint(currentMode))
+		if err != nil {
+			return nil, errDualWriterSetCurrentMode
+		}
+	}
+	// 	#TODO add support for other combinations of desired and current modes
```

Comment:
switching from mode 2 back to mode 1 should be similarly straightforward

Reply:
True. Will add this one as soon as we switch to using implicit feature flags and config. Once I can confirm the approach is correct, I'll add the necessary config to check if the desired mode is mode 1. Should be fine for now since we are currently defaulting to mode 1 unless the desired mode is mode 2.